packable/packable/src/packable/map.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright 2024 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Types and implementations for packing and unpacking map values.
+
+extern crate alloc;
+
+#[cfg(feature = "usize")]
+use core::borrow::Borrow;
+use core::{convert::Infallible, fmt};
+#[cfg(feature = "std")]
+use std::{collections::HashMap, hash::Hash};
+
+#[cfg(feature = "usize")]
+use crate::{error::UnpackError, packer::Packer, unpacker::Unpacker, Packable};
+
+/// Error type raised when a semantic error occurs while unpacking a map.
+pub enum UnpackMapError<K, KE, VE, P> {
+    /// A duplicate key.
+    DuplicateKey(K),
+    /// Semantic error raised while unpacking a key of the map. Typically this is
+    /// [`Packable::UnpackError`].
+    Key(KE),
+    /// Semantic error raised while unpacking a value of the map. Typically this is
+    /// [`Packable::UnpackError`].
+    Value(VE),
+    /// Semantic error raised when the length prefix cannot be unpacked.
+    Prefix(P),
+}
+
+impl<K, KE: fmt::Debug, VE: fmt::Debug, P: fmt::Debug> fmt::Debug for UnpackMapError<K, KE, VE, P> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::DuplicateKey(_) => f.debug_tuple("DuplicateKey").finish(),
+            Self::Key(arg0) => f.debug_tuple("Key").field(arg0).finish(),
+            Self::Value(arg0) => f.debug_tuple("Value").field(arg0).finish(),
+            Self::Prefix(arg0) => f.debug_tuple("Prefix").field(arg0).finish(),
```

Comment:
I would rather use `err` here like for `Display` instead of `arg0` tbh

Reply:
I just copy pasted some existing code but I guess I can change, yes

Reply:
Done